### PR TITLE
Refactor integration tests for clarity and speed

### PR DESCRIPTION
**Pull request by stevedlawrence**
The current integration tests have too much boilerplate that can be
refactored with some new APIs. Additionally, the integration tests are
incredibly slow due to all the forking and creation of new VMs. This
fixes both of these issues.

The new integration test API creates a new `runCLI` function, accepting
an array of string arguments (with a new "args" string interpolator to
make specifying this array easier), various execution parameters, a
function that contains the normal `Expect` logic all our of integration
tests already use, and finally the expected exit code.

A new `path` function is created to help deal with paths in a OS
agnostic manner.

New `withTempFile` and `withTempDirectory` functions are added to help
with creating temporary files in a special temporary daffodil directory,
and to ensure they are cleaned up when the test is complete.

A new `withSysProp` function is used to temporarily change and restore
system properties, helping to avoid changes that might interfere with
other tests.

With these new functions, integration tests now look something like
this:

```scala
    val schema = path("path/to/schema.dfdl.xsd")
    runCLI(args"parse -s $schema") { cli =>
      cli.send("data to parse", inputDone = true)
      cli.expect("<data>string to expect</data>")
    } (ExitCode.Success)
```

The runCLI function supports executing Daffodil in either a new process
or thread, defaulting to thread unless a test requires a separator
process (e.g. classpath changes). Streams are setup for handling
stdin/out/err. When threaded, the thread calls the
org.apache.daffodil.Main run function, and configures Main and log4j to
output to these streams. When using a new processing, the daffodil
sh/bat script is directly executed.

A new `Expect` instance is configured so that it can read/write to these
streams and validate the output, with a new CLITester class used to
provide helper wrappers around `Expect` to make logic a bit less
verbose.

Multiple changes were also made to Main and the Debugger to ensure they
always use the specified streams instead of always using stdin/out/err.
This is really only needed for this integration testing, but could
potentially be useful in other use cases where stdin/out/err need need
to be mimicked without forking a new process.

Note that the thread approach isn't thread safe, but integration tests
are already not run in parallel. By using threads instead of spawning
new processes where possible, we drastically decrease the overhead, with
test suites now taking a number of seconds instead of a number of
minutes, averaging and order of magnitude speed up.

This also tweaks various CLI tests to not expect data from files,
instead hardcoding what to expect in each test. This is a bit less
exact, but it make the logic simpler and more consistent, and allows us
to remove many files from the repo. Also found and removed many debugger
command files that are never actually used.

This also enables the blob CLI tests but adds an assumption so that they
are skipped if the large input files have not been generated. This gives
us a reminder that those tests exist and we can periodically ensure they
work by generating the files.

This also ensures coverage is enabled in the github workflow every time
we execute SBT. Otherwise, changes of coverage enablement cause a
recompile. By doing this we avoid recompiling the same files multiple
times which saves a couple of minutes overall. Note that we originally
disabled coverage on some steps because it caused failures, but whatever
caused that seems to have been fixed.

DAFFODIL-2381

### CODE REVIEW

**Review by tuxji — COMMENTED** [review]
I think the new API looks fine and I would like to see the rest of the integration tests converted to use it too.  The converted tests are much smaller as well as run much faster, which is a big quality of code improvement.  As I noted in a couple of places, we might even be able to remove the old API entirely (which I would prefer).  

We may need to pay attention to other places in Daffodil which call some variation of "println" and determine if these places need to be updated or not.  We also may need to increase code coverage (this pull request currently hits only 65% code coverage).

File: daffodil-cli/src/it/scala/org/apache/daffodil/parsing/TestCLIParsing.scala
Comment (stevedlawrence) [reply daffodil-cli/src/it/scala/org/apache/daffodil/parsing/TestCLIParsing.scala:615]: I've converted pretty much everything to this new API, you are right that tests that set DAFFODIL_JAVA_OPTS don't really need to do that, and in the case where it is needed there are workarounds using setProperty.

However, most of the UserDefinedFunction integration tests set the DAFFODIL_CLASSPATH variable. Essentially what these tests do is put all `.class` files defined in daffodil-udf on the classpath, and then individual tests also add specific `META-INF/services` directories on the classpath to tests loading specific UDF. I'm not sure there's a way to modify the classpath like this without forking a new process? Any ideas oh if that's possible?

Another consideration is that the `UserDefinedFunctionService` object loads all the available UDF's on initialization and stores their information in a mutable HashMap. So if we were able to have classpath specific for these tests, we would also need to ensure this object is reset/reloaded appropriately. It's possible sbt magic already handles resetting objects? I feel like they don't persist inbetween tests, or maybe test suites?

Any thoughts?

Note that even if we do fork for only the UDF integration tests, it still feels like a pretty big win. With the current changes, all integration tests take about 2 minutes on my machine (half of that is UDFs), which is a big improvement over the usual 20+ minutes.

Comment (tuxji) [reply daffodil-cli/src/it/scala/org/apache/daffodil/parsing/TestCLIParsing.scala:615]: Yes, it's possible but far from easy to search a different classpath at runtime without forking a new process.  Java 9 and later versions have removed the hackish (it was never part of the API) ability to modify the system classloader at runtime.  The only portable solution is to create a custom classloader and modify your application to use the custom classloader instead of the system classloader.  There appears to be no solution that will allow the UserDefinitionFuction integration tests to avoid forking unless we modify not only these tests, but also the daffodil-udf code (if it's possible to restrict the scope of the changes to only that library), or else the Main class itself.  I think we will have to be satisfied with the speedup we get from converting most of the integration tests and accept the slowness of forking for the UDF integration tests.  Running all integration tests in 2 minutes instead of 20+ minutes is still wonderful.

Here are the two best StackOverflow questions about changing the Java classpath at runtime.  I also googled for Scala as well as Java answers and didn't find a better approach although I did find a Scala library which lets you search any directory or jars for classes (the drawback is you can't load these classes unless the classes are also loadable by your system classloader or a custom classloader).

https://stackoverflow.com/questions/252893/how-do-you-change-the-classpath-within-java
https://stackoverflow.com/questions/60764/how-to-load-jar-files-dynamically-at-runtime
http://software.clapper.org/classutil/

**Review by mbeckerle — APPROVED** [review]
+1

This is a big improvement. The speed will be very nice, but also elimintes a bunch of portability concerns with the tests also where MS-Windows-isms were not adeuqately abstracted.

That is, this fixes DAFFODIL-2381

I belive this will also make debugging these tests more straightforward, as one can just set breakpoints in Daffodil per usual and debug the test, rather than various temp/hack ways Ih've had to modify the tests (have them call main, instead of spawning a process) for debugging in IDE.

**Comment by stevedlawrence** [comment]
I've completed the refactor using a new API and address all comments, I think this is ready for a full review.

**Comment by stevedlawrence** [comment]
Looks like the PR checks finished in 17-25, with most taking about 20 mins. Compared to the most recent merged commit which took between 30-60 averaging on the higher end, this is a nice speed up.

**Comment by mbeckerle** [comment]
This will be post 3.4.0, i.e., on the 3.5.0 branch yes?

**Comment by stevedlawrence** [comment]
> This will be post 3.4.0, i.e., on the 3.5.0 branch yes?

Correct, once this gets the +1's, I'll hold off merging until 3.4.0 is released.

**Review by mbeckerle — APPROVED** [review]
+1

File: daffodil-cli/src/it/scala/org/apache/daffodil/saving/TestCLISaveParser.scala
Comment (tuxji) [inline daffodil-cli/src/it/scala/org/apache/daffodil/saving/TestCLISaveParser.scala:36]: This PR inlines these files' expected strings into the tests below and uses withTempFile { parser instead of savedParserFile.  Nice!

File: daffodil-cli/src/it/scala/org/apache/daffodil/saving/TestCLISaveParser.scala
Comment (tuxji) [inline daffodil-cli/src/it/scala/org/apache/daffodil/saving/TestCLISaveParser.scala:217]: This PR uncomments the test's body and keeps only the `@Test` commented out.  Nice, this will keep the test compiling even after changes to functions.

File: daffodil-cli/src/it/scala/org/apache/daffodil/unparsing/TestCLIUnparsing.scala
Comment (tuxji) [inline daffodil-cli/src/it/scala/org/apache/daffodil/unparsing/TestCLIUnparsing.scala:171]: Again, nice job keeping a commented out test compiling.

File: project/Rat.scala
Comment (tuxji) [inline project/Rat.scala:57]: Does the PR delete these files because their contents had already been inlined as input strings into the corresponding CLI tests in an earlier PR?  Nothing seems to read or use these files even before this PR was created, so it's time to remove them.

**Review by tuxji — APPROVED** [review]
+1

File: daffodil-cli/src/it/scala/org/apache/daffodil/CLI/Util.scala
Comment (stevedlawrence) [reply daffodil-cli/src/it/scala/org/apache/daffodil/CLI/Util.scala:278]: I used this command to find non-test related println/printf:

    grep -R --exclude-dir=test -e '\(println\|printf\)' * 

This found the following:

* Main/CLIDebuggerRunner/TraceDebuggerRunner - these all use new logic to output to a specified stream
* DirectOrBufferedDataOutputStream.scala - this println occurs in the `dumpChain()` function, which is used for debugging suspensions
* PropertyGenerator.scala - run by sbt to generate code--needs to always output to stdout for sbt
* CodeGenerator.scala - run by sbt to update example files--needs to be output to stdout for sbt
* Tak.scala - used for timing TDML tests/performance related

All other instances are in commented out code. So I think all instances that related to the CLI have been fixed, or are intended to output to stdout. We might want to go through all the tests and commented out code and inspect those println's to see if they are really needed, but I think that's out of scope for this ticket.

File: daffodil-cli/src/it/scala/org/apache/daffodil/executing/TestCLIexecuting.scala
Comment (stevedlawrence) [reply daffodil-cli/src/it/scala/org/apache/daffodil/executing/TestCLIexecuting.scala:34]: The `TestCLIexecuting` and `TestCLIlisting` files were both really about running the CLI "test" subcommand to run TDML tests, with one focused on executing tdml tests and the other focused on listing tdml tests. And in fact, the "executing" file even had multiple tests that were just listing tests, and I found some tests that were duplicated in the two files, so the intended distinction wasn't even clear. I also think "executing" also very clear since that could be about executing anything, like  parse, unparse, code generator, etc.

So to make it more obvious that this is the file for the "test" command, I combined these two files into a "TestCLItdml" file (I thought "TestCLItest" might be confusing?) and removed duplicates when I found them.

File: project/Rat.scala
Comment (stevedlawrence) [reply project/Rat.scala:57]: Looking at the commit history, when debugger tests were originally added (nearly 10 years) ago they used the `-d` option and passed in these files. A short time later (04e50f9aad93eae15a02d7eff4b9440e84bc9e2d), we switched to ExpectIt and changed these tests to inline the debugger commands, and I guess we just forgot to remove the files.

**Comment by stevedlawrence** [comment]
> We also may need to increase code coverage (this pull request currently hits only 65% code coverage).

I didn't really add any new code except in couple minor places. I think lack of coverage in this PR is because I've made changes to printlns that were just already missing coverage.

In the  past I know I've hesitated to add new CLI tests to get coverage because of the high overhead. But now that integration tests are much quicker, we may want to consider adding new tests to improve our CLI coverage. I think that's out of scope for this PR though, since this is more about just improving the CLI test infrastructure.

**Comment by mbeckerle** [comment]
When this says 64% coverage, is that the unit tests plus the CLI tests? Or just the CLI tests alone?

**Comment by stevedlawrence** [comment]
I believe it's both unit tests + CLI tests. But most of the lines missing coverage are println's in Main.scala that can only be tested with he CLI/Integration tests.

**Comment by mbeckerle** [comment]
Something doesn't add up.

We were at roughly 80% coverage on a 110K-lines code base.

We dropped to 64% coverage.  That's a 16% drop, meaning 17K-lines are no longer covered that were.

That's can't be explained by println statements in the Main/CLI code. 

So I'm assuming my arithmetic here doesn't make sense. But I clearly don't understand what we mean by 80+% dropping to 64% coverage.

**Comment by stevedlawrence** [comment]
I think that 64% is just the the coverage of this PR, not the coverage of the entire codebase. So of the lines this PR changed to `src/main`, 64% of those lines have a test that hit them. The overall coverage is still about 80%.

Here's the codecov page for this PR that makes it a bit more clear:

https://app.codecov.io/gh/apache/daffodil/pull/865

**Comment by mbeckerle** [comment]
Ok. Then I suggest adding a ticket to enhance code coverage here. But I wouldn't hold up committing this much improvement for that.

**Comment by stevedlawrence** [comment]
[DAFFODIL-2746](https://issues.apache.org/jira/browse/DAFFODIL-2746) added to improve CLI code coverage.

**Pull request merged**
